include/zephyr/drivers/pinctrl/pinctrl_nrf_common.h

```diff
@@ -0,0 +1,259 @@
+
+/** Struct describing nrf_gpio configuration to be applied. */
+struct nrf_pinctrl_config {
+	enum nrf_pinctrl_write write;
+	nrf_gpio_pin_dir_t dir;
+	nrf_gpio_pin_input_t input;
+	nrf_gpio_pin_drive_t drive;
+	nrf_gpio_pin_sense_t sense;
+	bool clock_pin_en;
+	uint8_t pull;
+};
+
+/**
+ * @brief Utility macro to check if a function requires clockpin enable.
+ *
+ * @param node_id Node identifier.
+ * @param prop Property name.
+ * @param idx Property entry index.
+ * @param p_node_id Parent node identifier.
+ */
+#define Z_CHECK_CLOCKPIN_ENABLE(node_id, prop, idx, fun) \
+	DT_PROP_BY_IDX(node_id, prop, idx) == fun ? BIT(NRF_CLOCKPIN_ENABLE_POS) :
+
+/**
+ * @brief Utility macro compute the clockpin enable bit.
+ *
+ * @note DT_FOREACH_PROP_ELEM_SEP_VARGS() is used instead of
+ * DT_FOREACH_PROP_ELEM_VARGS() because the latter is already resolved in the
+ * same run.
+ *
+ * @param node_id Node identifier.
+ * @param prop Property name.
+ * @param idx Property entry index.
+ * @param p_node_id Parent node identifier.
+ */
+#define Z_GET_CLOCKPIN_ENABLE(node_id, prop, idx, p_node_id)		            \
+	COND_CODE_1(DT_NODE_HAS_PROP(p_node_id, nordic_clockpin_enable),            \
+		    ((DT_FOREACH_PROP_ELEM_SEP_VARGS(			            \
+			p_node_id, nordic_clockpin_enable, Z_CHECK_CLOCKPIN_ENABLE, \
+			(), NRF_GET_FUN(DT_PROP_BY_IDX(node_id, prop, idx)))        \
+		      0)), (0))
+
+/**
+ * @brief Utility macro to get the GPD_FAST_ACTIVE1 flag
+ *
+ * @param p_node_id Parent node identifier.
+ */
+#define Z_GET_GPD_FAST_ACTIVE1(p_node_id)				       \
+	COND_CODE_1(DT_NODE_HAS_PROP(p_node_id, power_domains),		       \
+		    ((DT_PHA(p_node_id, power_domains, id) ==		       \
+		      NRF_GPD_FAST_ACTIVE1) << NRF_GPD_FAST_ACTIVE1_POS), (0))
+
+/**
+ * @brief Utility macro to initialize each pin.
+ *
+ * @param node_id Node identifier.
+ * @param prop Property name.
+ * @param idx Property entry index.
+ * @param p_node_id Parent node identifier.
+ */
+#define Z_PINCTRL_STATE_PIN_INIT(node_id, prop, idx, p_node_id)		       \
+	(DT_PROP_BY_IDX(node_id, prop, idx) |				       \
+	 ((NRF_PULL_DOWN * DT_PROP(node_id, bias_pull_down)) << NRF_PULL_POS) |\
+	 ((NRF_PULL_UP * DT_PROP(node_id, bias_pull_up)) << NRF_PULL_POS) |    \
+	 (DT_PROP(node_id, nordic_drive_mode) << NRF_DRIVE_POS) |	       \
+	 ((NRF_LP_ENABLE * DT_PROP(node_id, low_power_enable)) << NRF_LP_POS) |\
+	 (DT_PROP(node_id, nordic_invert) << NRF_INVERT_POS) |		       \
+	 Z_GET_CLOCKPIN_ENABLE(node_id, prop, idx, p_node_id) |		       \
+	 Z_GET_GPD_FAST_ACTIVE1(p_node_id)				       \
+	),
+
+/**
+ * @brief Utility macro to initialize state pins contained in a given property.
+ *
+ * @param node_id Node identifier.
+ * @param prop Property name describing state pins.
+ */
+#define Z_PINCTRL_STATE_PINS_INIT(node_id, prop)			       \
+	{DT_FOREACH_CHILD_VARGS(DT_PHANDLE(node_id, prop),		       \
+				DT_FOREACH_PROP_ELEM_VARGS, psels,	       \
+				Z_PINCTRL_STATE_PIN_INIT, node_id)}
```

Comment:
Why is all this stuff from `<pinctrl_soc.h>` moved here? In my opinion, we should have here only the hook declarations, `struct nrf_pinctrl_config`, and `NRF_GET_*` helpers (I don't see the need to introduce `struct nrf_pinctrl_pin`, hooks can be provided with the `pincfg` and `reg` values instead).

Reply:
To keep everything in one place. This is in line with how other vendors do it:
https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/drivers/pinctrl/pinctrl_rcar_common.h#L100